쥬스메이커 [STEP 2] 예톤, Hugh

Hugh-github opened the pull request:
안녕하세요 제이슨!@ehgud0670 
Step02 PR 제출합니다.

## step2 기능 구현
- ViewController
    - 재고수정 버튼을 눌렀을 때 화면이 전환되는 메서드(tabEditStock)
    - 쥬스 주문 버튼을 눌렀을 때 쥬스를 주문하는 메서드(order)
    - 과일들의 재고(label) 수량을 변경시켜주는 메서드(updateStock)
    - 재고 부족 알림창을 띄어주는 메서드(showlackOfStockAlert) 
    - 쥬스 제조 완료 알림창을 띄어주는 메서드(showCompletionAlert)
    - 재고 수정 화면으로 넘어가는 방식을 modality로 구현한 메서드(changeView)
        - 메서드(showlackOfStockAlert)의 okAlert의 handler에서 사용
        - 재고수정버튼 눌렀을 때 사용
- JuiceMaker
- FruitStore
    - 원래 기본값으로 설정했던 초기재고를 initializer를 통해 초기화해주는 방식으로 변경
    - stock에 저장돼있는 과일의 개수를 ViewController으로 가져다주는 메서드(notifyStock) 구현
- Menu
    - Menu의 case를 String 타입으로 반환해주는 계산 프로퍼티 name 변수 선언
 
## 고민했던 것들
- 네비게이션 컨트롤러가 두 개인 이유에 대해 고민
    - 왜 이번 프로젝트에서 네비게이션 컨트롤러가 2개 있는지에 대해 토론해봤습니다. 결론은 나중에 앱이 확장될 것을 생각하면 재고 수정만의 문제가 아니라 훨씬 더 많은 화면들이 필요할텐데 하나의 네비게이션 컨트롤러에 연결된 흐름으로 재고수정 화면을 구현해주면 나중돼서 다 바꿔줘야할 수도 있어 번거로울 것이다 라는 결론이 나왔습니다. 네비게이션 컨트롤러는 스택을 담는 바구니라고 생각했습니다. 저희는 재고 수정 화면과 과일 주문 화면의 흐름이 다르다고 결론이 나서 모달로 화면을 구성하기로 했습니다. 또 그에따라 화면이 하나의 흐름이 아닌 두 갈래의 흐름이 되기 때문에(현재로썬) 두 개의 네비게이션 컨트롤러를 사용해주기로 결정했습니다. 저희가 생각한 결론이 맞을까요?
- 과일쥬스 주문 버튼을 눌렀을 때 그걸 Menu의 case로 인식하기 위한 방법 2가지 중 뭘 쓸지 고민했습니다.
    - 1번째 방법: currentTitle(버튼에 쓰여진 문자열을 그대로 인식) 사용
    - 2번째 방법: 각 주문 버튼에 tag를 설정해주고 enum Menu의 원시값을 Int형으로 해주고 둘의 값을 비교해줌
    - 저희는 tag를 사용하는 방법이 깔끔해보여서 사용했습니다. 혹시 더 좋은 방법이 있을까요? 
- 에러처리를 ViewController의 order 메서드에서 해줘야 할 것 같아 코드를 수정했습니다. 저희가 수정한 방법이 맞을까요?
  - Step01에서는 FruitStore의 decreaseStock에서 에러를 던지고 JuiceMaker의 make 메서드에서 에러를 처리해주었습니다.
  - JuiceMaker의 make함수에서 에러를 던지고 ViewController의 order 메서드에서 에러를 처리해주었습니다.
```swift
// JuiceMaker
    func make(menu: Menu) throws {
        guard store.hasFruit(menu: menu) else {
            throw JuiceMakerError.lackOfStock
        }
        store.decreaseStock(menu: menu)
    }

// ViewController
@IBAction func order(_ sender: UIButton) {
        guard let juice = Menu(rawValue: sender.tag) else {
            return
        }
        do {
            try juiceMaker.make(menu: juice)
        } catch {
            showlackOfStockAlert()
        }
        updateStock()
        showCompletionAlert(menu: juice)
    }
```

Hugh-github committed:
feat: 버튼 누르면 쥬스를 주문하는 함수 선언

- tag와 주문 버튼을 연결
- JuiceMaker, FruitStore, Menu에 존재하는 함수의 파라미터(total) 제거

yeeton37 committed:
feat: 현재 재고의 변화를 감지해서 재고 label의 값을 변경해주는 함수 선언

- updateStock: 과일 재고 라벨이 업데이트 됨
- notifyStock: FruitStore의 stock을 넘겨줌
- JuiceMaker안에서 FruitStore의 인스턴스를 담는 변수를 let으로 변경 및 접근 제어 수준 변경

Hugh-github committed:
feat: 재고가 부족하거나 쥬스가 완성되면 뜨는 알림 메시지 출력하는 함수 선언 및 에러 처리 수정

- ViewController에서 에러처리가 가능하도록 코드를 수정

yeeton37 committed:
feat: 화면 전환하는 메서드(changeView)선언 및 Menu 타입 확장

- Menu타입을 확장해서 만든 속성을 메서드(showCompletionAlert)에 활용
- 화면 전환하는 메서드를 선언해 재고수정 버튼과 연결, 재고부족 알림의 handler와 연결

Hugh-github committed:
feat: 초기재고 저장하는 방식 변경

- property 기본값을 제거하고 initializer를 통해 초기화 해주는 방식으로 변경

yeeton37 committed:
docs: README.md 수정

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -5,38 +5,50 @@
     
-    private func hasFruit(menu: Menu, total: Int) -> Bool {
-        let necessaryFruit = menu.count(to: total)
+    init(defaultStock: Int) {
+        Fruit.allCases.forEach {
+            stock[$0] = defaultStock
+        }
+    }
```
```suggestion
    init(need: Int = 10) {
        Fruit.allCases.forEach {
            stock[$0] = defaultStock
        }
    }
```
=> 저번에 말씀드렸던건 위처럼 Default Parameter Value 를 집어넣으라는 말씀이었습니다.
그럼 평소에는 FruitStore()로 생성하다가도 테스트할 때 임의의 need 값을 집어넣을 수 있습니다 :) 

[해당 페이지](https://docs.swift.org/swift-book/LanguageGuide/Functions.html)에서 Default Parameter Values 로 검색해서 어떤것인지 알아보세요!

ehgud0670 replied:
음 다시 보니 need라는 이름은 부적절하네요. amount?

Hugh-github replied:
Defualt Parameter Value를 이용해 수정했습니다.
제이슨의 말을 이제 이해했네요 😆

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -5,38 +5,50 @@
+    init(defaultStock: Int) {
+        Fruit.allCases.forEach {
+            stock[$0] = defaultStock
+        }
```
allCass를 쓰신것은 매우 잘하셨네요! 다만 Int 값인 amount(임의상 amount로 말씀드렸습니다)만 init으로  주입받기 보단 stock 자체(type: `[Fruit: Int]`) 를 주입받는 건 어떨까요? 좀 더 유연하게 데이터를 집어넣을 것 같네요!

Hugh-github replied:
```swift
class FruitStore {
    private var stock: [Fruit: Int]
    
    init(stock: [Fruit: Int]) {
        self.stock = stock
    }
}
```
위 코드처럼 작성하는게 맞을 까요? 
위 코드 처럼 작성하면 필요한 과일과 수량만 입력 받을 수 있어 제이슨의 말대로 유연하게 데이터 입력이 가능할 것 같습니다.

ehgud0670 replied:
네 좋습니다 :)

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -5,14 +5,12 @@
 
 struct JuiceMaker {
-    private var store = FruitStore()
+    let store = FruitStore(defaultStock: 10)
```
여기 배우신 것처럼 private 으로 접근 제한하시면 될 듯 합니다. 다른 곳도 안되어있는 곳 모두 private으로 해주세요!

Hugh-github replied:
모든 메서드와 속성을 확인하고 접근제어 수준을 변경했습니다!!

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/Menu.swift:
```diff
@@ -33,13 +33,33 @@ enum Menu {
     
-    func count(to total: Int) -> [Fruit: Int] {
-        var countList: [Fruit: Int] = [:]
+    func count() -> [Fruit: Int] {
```
그냥 count 라고 하니깐 Int를 반환하는 것 같아서 파악이 바로 안되네요~
`needsForFruits()` 라고 하는건 어떨까요? 
저의 제안이고 더 좋은 네이밍이 떠오르시면 다른 걸로 바꾸시면 될 듯 합니다 :)

Hugh-github replied:
`count`-> `needsForFruits` 으로 변경했습니다.

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/Menu.swift:
```diff
@@ -33,13 +33,33 @@ enum Menu {
     }
 }
 
+extension Menu {
+    var name: String {
+        switch self {
+        case .strawberryJuice:
+            return "딸기쥬스"
+        case .mangoJuice:
+            return "망고쥬스"
+        case .kiwiJuice:
+            return "키위쥬스"
+        case .pineappleJuice:
+            return "파인애플쥬스"
+        case .bananaJuice:
+            return "바나나쥬스"
+        case .mangoAndKiwiJuice:
+            return "망키쥬스"
+        case .strawberryAndBananaJuice:
+            return "딸바쥬스"
+        }
+    }
```
네 이렇게 메뉴당 name 만드신거 잘하셨네요! 👍

Hugh-github replied:
감사합니다!!!! 🥳

ehgud0670 commented on README.md:
```diff
@@ -127,5 +127,93 @@ func classifyKey(menu: Menu) -> [Fruits] {
+     case strawberryJuice
+     case bananaJuice
+
+     private var recipe: [Fruit: Int] { // private 붙이고
```
사실 이건 일부러 private 을 두라고 말씀드렸던거고, 원래는 연산 프로퍼티는 internal로 접근 제한 줘도 됩니다. 
참고차 말씀드려요!

ehgud0670 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,12 +7,71 @@
 import UIKit
 
 class ViewController: UIViewController {
```
VC 도 그냥 VC가 아니라 이름을 붙여줍니다. 코어가 되는 VC는 HomeViewController 라든가 이름을 붙이곤 합니다 :)

Hugh-github replied:
네 ViewController 파일 이름을 HomeViewController로 변경했습니다!

ehgud0670 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,12 +7,71 @@
+        present(lackOfStockAlert, animated: true)
+    }
+    
+    func showCompletionAlert(menu: Menu) {
+        let completionAlert = UIAlertController(title: "알림", message: "\(menu.name) 나왔습니다! 맛있게 드세요!", preferredStyle: .alert)
+        let okAlert = UIAlertAction(title: "OK", style: .default, handler: nil)
+        
+        completionAlert.addAction(okAlert)
+        
+        present(completionAlert, animated: true)
+    }
+    
```
여기 다 private 붙여줘야 겠네요!

ehgud0670 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,12 +7,71 @@
+        showCompletionAlert(menu: juice)
+    }
+    
+    func updateStock() {
```
updateStock이라고 하니깐 모델인 stock 값을 업데이팅하는것 같아요. 
Label들의 text 값을 업데이팅한다는 것이 메소드 명에 좀 더 명확히 보이면 좋을 것 같습니다 :)

Hugh-github replied:
확실히 Label이나 text가 없어 함수의 기능을 바로 파악하는 게 어려운 것 같아 수정했습니다.

ehgud0670 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,12 +7,71 @@
+    }
+    
+    func updateStock() {
+        strawberryStock.text = juiceMaker.store.notifyStock(fruit: .strawberry).description
```
notifyStock이 Int를 뱉는데 결국 필요한 값은 String 이니깐 description을 사용하시는 것 같은데, 아래처럼 직접적으로 변환하는게 좋을 것 같아요. 
```suggestion
        strawberryStock.text = "\(juiceMaker.store.notifyStock(fruit: .strawberry))"
```
왜냐하면 description은 Int 말고도 다른 어떤 타입도 가지고 있는 값이라 description을 쓰면 어떤 텍스트 값인지 바로 보이지 않지만, 위처럼 쓰면 더 분명하게 text값으로 어떤값이 할당되는지 잘 보여서입니다.

description과 관련된 Customstringconvertible protocol도 알고 계시면 좋습니다.
https://developer.apple.com/documentation/swift/customstringconvertible

yeeton37 replied:
description을 쓸 때보다 문자열 보간법처럼 사용해주니까 확실히 Int-> String 타입으로 변환해주고자 하는 목적이 확실히 보이네요! 변경해주었습니다 🚀

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -5,38 +5,50 @@
+        }
+    }
+
+    func notifyStock(fruit: Fruit) -> Int {
```
왜 notify라고 하는지 알것 같긴 하지만, notify 네이밍는 주로 NoficationCenter.default.post를 쓸때 네이밍을 써서 헷갈리긴 하네요~
swift 에서는 보통 멤버 프로퍼티를 반환할때 해당 프로퍼티 이름 그대로 사용하는 네이밍을 사용하는데 명사형으로 네이밍하시는것 어떠실까요?
https://github.com/StyleShare/swift-style-guide#%ED%95%A8%EC%88%98

ehgud0670 replied:
좋은 예:
```swift
func name(for user: User) -> String?
```
나쁜 예:
```swift
func getName(for user: User) -> String?
```

Hugh-github replied:
함수를 네이밍을 할 때 동사형으로 네이밍을 했는데 멤버 프로퍼티를 반환하는 함수는 프로퍼티 이름을 그대로 함수의 네임으로 사용할 수 있다는 걸 알려주셔서 감사합니다! 🥹

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/Menu.swift:
```diff
@@ -5,7 +5,7 @@
 //
 
-enum Menu {
+enum Menu: Int {
```
순서를 정해주기 위해 Int를 채택하는 건 나쁜 패턴입니다. 
순서를 정해야 한다면 order를 반환하는 연산 프로퍼티를 만드시는게 좋습니다.

yeeton37 replied:
아 그런줄 몰랐네요...!
쥬스주문 버튼의 tag부분과 Menu의 Int부분을 모두 지우고 쥬스 주문 버튼을 뷰컨트롤러의 IBOutlet에 연결해주었습니다!

ehgud0670 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,12 +7,71 @@
+        mangoStock.text = juiceMaker.store.notifyStock(fruit: .mango).description
+    }
+    
+    func showlackOfStockAlert() {
```
여기 카멜케이스 잘못 되어있네요!

Hugh-github replied:
showLackOfStockAlert  수정했습니다!

ehgud0670 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,12 +7,71 @@
-
+    var juiceMaker = JuiceMaker()
+    
+    @IBOutlet weak var strawberryStock: UILabel!
```
```suggestion
    @IBOutlet weak var strawberryStockLabel: UILabel!
```
끝에 Label이라고 명시를 해야 좀더 변수가 분명해보이네요 😄

Hugh-github replied:
Label을 추가해 수정했습니다!

ehgud0670 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,12 +7,71 @@
+        present(completionAlert, animated: true)
+    }
+    
+    func changeView() {
```
뷰를 실질적으로 변화시킨건 아니고 다른 화면을 띄우는 메소드였네요
그럼 presentEditStockViewController 처럼 화면을 띄우는 메소드 네이밍이 더 적절해보입니다.

Hugh-github replied:
모달을 이용해 다른 화면을 띄우는 기능을 하고 있으니 제이슨의 말처럼 presentEditStockViewController로 변경했습니다.

ehgud0670 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,12 +7,71 @@
+    }
+    
+    func changeView() {
+        guard let stockVC = self.storyboard?.instantiateViewController(withIdentifier: "EditStockViewController")
```
stockVC가 아니라 editStockVC 네요.

Hugh-github replied:
editStockVC로 변경했습니다.

ehgud0670 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,12 +7,71 @@
+        self.changeView()
+    }
+    
+    @IBAction func order(_ sender: UIButton) {
```
1. tag를 쓰지말고 button들을 IBOutlet으로 직접 VC에 선언해서 사용하는게 좋습니다. 
현재 Menu의 rawValue를 순서값으로 사용하고 해당 순서값을 그대로 tag값으로 사용해서 코드를 구현하는데 좋지 않은 패턴이라서요!

2. 참고차 말씀드리면 tag는 동적으로 생성된 뷰를 취급할 때 방법이 없을때 사용하기 좋은 값이지, 지금처럼 이미 화면에 보이는 버튼들에 사용하기 좋은 방법은 아닙니다. 
tag를 사용하면 앱은 tag를 통해 런타임적으로 어떤 뷰인지 판별합니다. 
즉 이미 알고 있는 뷰라면 tag를 쓰는 것이 비효율적이라는 거죠.

yeeton37 replied:
위에 댓글 단 것과 같이 tag를 삭제해주고 버튼을 VC와 직접 연결해서 주문을 할 수 있도록 해주었습니다.
주문 하는 메서드 안에서 버튼을 쥬스타입으로 변환해주기 위해 convertToJuice 메서드를 따로 선언해주었습니다!

ehgud0670 commented on JuiceMaker/JuiceMaker/View/Base.lproj/Main.storyboard:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="20037" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES" useSafeAreas="YES" colorMatched="YES" initialViewController="DCG-dP-Fms">
-    <device id="retina6_0" orientation="landscape" appearance="light"/>
```
스토리보드도 여러개 만들수 있습니다. 
현재 뷰컨이 둘이 있는데 각각의 스토리보드를 만드는건 어떨까요?
보통 협업할때 스토리보드 한곳에서 여러 화면을 만들면 충돌나기 쉽기때문에 여러 스토리보드로 나누곤합니다.
지금은 ViewController가 2개밖에 없긴하지만 시험삼아 나눠보시는건 어떨까요 ㅎㅎ

ehgud0670 reviewed (changes requested):
많은 걸 배우셨네요! 고생하셨습니다.
피드백 확인 부탁드릴게요 🙏

ehgud0670 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,12 +7,71 @@
+    
+    func updateStock() {
+        strawberryStock.text = juiceMaker.store.notifyStock(fruit: .strawberry).description
+        bananaStock.text = juiceMaker.store.notifyStock(fruit: .banana).description
+        pineappleStock.text = juiceMaker.store.notifyStock(fruit: .pineapple).description
+        kiwiStock.text = juiceMaker.store.notifyStock(fruit: .kiwi).description
+        mangoStock.text = juiceMaker.store.notifyStock(fruit: .mango).description
```
여기 레이블들도 연산프로퍼티로 배열화해서 텍스트값 할당하는 코드를 처리하면 코드 수 줄어들겠네요! 
DRY(Don't Repeat Yourself) 원칙을 고수하십시오!

Hugh-github replied:
FruitStockLabel 클래스를 상속받는 클래스를 만들어 코드를 수정했습니다.
push한 파일에서는 메서드 convertToFruit을 재정의해 사용하지만 아래와 같이 연산 프로퍼티를 이용해 하는 방법도 고민해봤습니다.
```swift
class FruitStockLabel: UILabel {
    var fruit: Fruit?
}

class StrawberryLabel: FruitStockLabel {
    override var fruit: Fruit? {
        get {
            return .strawberry
        } set(newFruit) {
            self.fruit = newFruit
        }
    }
}
```

ehgud0670 replied:
네네 get 만 하는게 더 좋아보이네요. 왜 그런지 고민해보세요 😄

yeeton37 committed:
refactor: 메서드 및 프로퍼티 네이밍 수정 및 이니셜라이저 default value로 수정

Hugh-github committed:
feat: IBOutlet(UIButton)을 사용해 쥬스를 주문하는 메서드(order) 수정, Menu 타입에서 CustomStringConvertible 채택

- order 메서드 수정 (tag를 사용하지 않고 UIButton과 Menu를 연결해 주는 메서드 생성(convertToJuice))
- convertToJuice에서 default일 경우 에러를 던지기 위해 JuiceMakerError에서 case(noneExistMenu) 추가

Hugh-github committed:
refactor: IBAction 대신 addTarget을 이용해 버튼 이벤트 처리

Hugh-github committed:
feat: label을 Fruit으로 변환해주는 메서드(convertToFruit) 선언

- label 을 연산 프로퍼티를 이용하여 배열화 및 메서드 updateFruitLabel에서 forEach를 사용해 label text에 문자열 할당

Hugh-github committed:
feat: UILabel을 관리하는 파일 FruitStockLabel 생성

yeeton37 committed:
feat: 각 과일 라벨 타입 선언 및 과일 재고 전달 메서드(transferFruitStock) 선언

yeeton37 committed:
refactor: 접근제어수준 변경 및 메서드(updateFruitStockLabel) 수정 및 노티피케이션센터 이름 확장

- 메서드(updateFruitStockLabel)에서 각 과일라벨 타입안의 메서드(convertToFruit)을 실행해 과일 재고 라벨 text를 변경해줌
- Notification.Name을 extension을 통해 notifyStock으로 바꿔줌

Hugh-github committed:
refactor: 함수명 변경, init의 default value 10으로 변경, 옵셔널 바인딩

ehgud0670 commented on JuiceMaker/JuiceMaker/Controller/HomeViewController.swift:
```diff
@@ -0,0 +1,120 @@
+        case mangoJuiceOrderButton:
+            return .mangoJuice
+        default:
+            throw JuiceMakerError.noneExistMenu
```
좋습니다 👍

ehgud0670 commented on JuiceMaker/JuiceMaker/Controller/HomeViewController.swift:
```diff
@@ -0,0 +1,120 @@
+import UIKit
+
+class HomeViewController: UIViewController {
+    var juiceMaker = JuiceMaker()
```
private 으로 아직 안해놓으셨네요! 
다음에 부탁드려요~

ehgud0670 commented on JuiceMaker/JuiceMaker/Controller/HomeViewController.swift:
```diff
@@ -0,0 +1,120 @@
+    
+    @IBOutlet var fruitStockLabel: [FruitStockLabel]!
+    
+    @IBOutlet weak var strawberryAndBananaJuiceOrderButton: UIButton!
+    @IBOutlet weak var mangoAndKiwiJuiceOrderButton: UIButton!
+    @IBOutlet weak var strawberryJuiceOrderButton: UIButton!
+    @IBOutlet weak var bananaJuiceOrderButton: UIButton!
+    @IBOutlet weak var pineappleJuiceOrderButton: UIButton!
+    @IBOutlet weak var kiwiJuiceOrderButton: UIButton!
+    @IBOutlet weak var mangoJuiceOrderButton: UIButton!
```
여기도 외부에서 참조하지 않는 거라면 private 으로 두시는게 좋을 것 같아요~

ehgud0670 commented on JuiceMaker/JuiceMaker/Controller/HomeViewController.swift:
```diff
@@ -0,0 +1,120 @@
+    @IBOutlet weak var mangoJuiceOrderButton: UIButton!
+    
+    var orderButton: [UIButton] {
+        return [strawberryJuiceOrderButton, mangoAndKiwiJuiceOrderButton, strawberryAndBananaJuiceOrderButton, bananaJuiceOrderButton, pineappleJuiceOrderButton, kiwiJuiceOrderButton, mangoJuiceOrderButton]
```
여기 행 줄이 너무 기네요! 
다음에 수정 부탁드려요~
스타일 가이드를 하나 참고하시는 것도 좋아요
저는 한국어로 되어 있는 [스타일 쉐어](https://github.com/StyleShare/swift-style-guide)의 스타일 가이드가 읽기 쉽고 이해하기 좋더라고요. 참고차 말씀 드립니다 ^^

ehgud0670 commented on JuiceMaker/JuiceMaker/Controller/HomeViewController.swift:
```diff
@@ -0,0 +1,120 @@
+        
+        NotificationCenter.default.addObserver(self, selector: #selector(updateFruitStockLabel), name: .notifyStock, object: nil)
+        
+        juiceMaker.transferFruitStock()
```
사실 stock의 변화는 뷰컨트롤러에서 juiceMaker나 fruitStore의 메소드를 직접 호출해서 알아내는게 아니라 
stock의 변화가 있을때, juiceMaker나 fruitStore가 알아서 노티를 보내고, vc는 관찰만 하고 있다가 노티로 변화를 감지하여 뷰를 업데이팅 하는 것입니다. 그리고 노티를 날릴때도 변화한 fruit들만 Notification에 보내도 됩니다.
그럼 참고하시고 고민하셔서 꼭 다음에 바뀌었으면 좋겠네요! 😃

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,39 +4,56 @@
     
-    private func hasFruit(menu: Menu, total: Int) -> Bool {
-        let necessaryFruit = menu.count(to: total)
+    func transferFruitStockToJuiceMaker() {
```
노티피케이션을 처음 써보시는데 잘 사용하셨어요! 😁 💯💯 🎉🎉🎉
transfer 라고 하기 보다는 notify나 post 가 메소드 이름에 더 어울리네요! ㅎㅎㅎ

ehgud0670 reviewed (approved):
1주간 고생많으셨습니다 ㅎㅎㅎ 
다시 피드백 드렸는데 다음에 반영하시면 좋을 것 같아요!!
그럼 월요일부터 1 주간 화이팅입니다 💪💪💪
화이팅 오토레이아웃!
